### DISCUSSION TITLE
Data compression of headers

### DISCUSSION THREAD

**Discussion by SunFulong** [wamp-proto/wamp-proto#538] [Ideas]
Do we have plan to compress the headers data like HTTP/2?

When announcing peer features, do authentication, passthrough or e2ee, option/detail filed name may be an overhead in the payload.

So like HTTP/2, if message using binary encoding, if we maintain a HARD CODE TABLE to map each filed name to no more than 2 bytes (65536 possible), will save the overhead.

e.g. the original hello message:

```
    [1, "somerealm", {
      "roles": {
          "publisher": {},
          "subscriber": {}
      }
    }]
```

we define: `0xx01` map to `roles`, `0x01 0x01` map to `publisher`, and `0x01 0x02` map to `subscriber`, when in binary encoding, this can save about 19 bytes, and also make the binary packet more binary

**Comment by oberstet** [comment]
> Do we have plan to compress the headers data like HTTP/2?

I would argue, nope - by design. But let's discuss of course, it's general design question relevant to WAMP.

Header compression is one way to compress (portions of) application payload.

Application payload compression, similar to application payload serialization or message framing, is a "WAMP transport level", definitely not "WAMP session level" feature.

Defining WAMP header value maps would follow a "WAMP session level" approach (thus, wrong design layer)

Further, WAMP already supports more powerful (general) compression of application payload, e.g. using WebSocket compression (deflate) - => fully supported in Crossbar.io and various Autobahns ..

**Comment by SunFulong** [comment]
> is a "WAMP transport level", definitely not "WAMP session level" feature.

Ya, I agree with you.

So, maybe we can discuss what if using RawSocket transport

**Comment by oberstet** [comment]
> So, maybe we can discuss what if using RawSocket transport

in the past, I have argued like this: the whole reason for the existence of RawSocket - given that there was WebSocket transport before already - has always been: *simplicity* in implementation and *high performance / minimal overhead*.

RawSocket fits that bill.

if you want "maximum features" - like for example application payload compression -, then looking for WebSocket as transport rather than gradually adding more feature to RawSocket - and thus risking the "simplicity" goal - does seem reasonable

btw: Autobahn and Crossbar.io also support non-standard WebSocket compression like not only Deflate based, but also Bzip2 based! And more .. I forgot .. ;)

**Reply by SunFulong** [reply]
Ok, I fully got your mean.

Thanks for your explain.

**Comment by oberstet** [comment]
right, Autobahn also supports WebSocket compression using [snappy](https://google.github.io/snappy/)!

https://github.com/crossbario/autobahn-python/blob/master/autobahn/websocket/compress.py
